src/auth/mod.rs

```diff
@@ -0,0 +1,202 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, you can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Types for parsing and authenticating HAWK headers.
+//! Matches the [Python logic](https://github.com/mozilla-services/tokenlib).
+//! We may want to extract this to its own repo/crate in due course.
+
+#[cfg(test)]
+mod test;
+
+use std::{
+    error::Error,
+    fmt::{self, Display, Formatter},
+};
+
+use actix_web::{error::ResponseError, http::header::ToStrError, FromRequest, HttpRequest};
+use base64::{self, DecodeError};
+use chrono::offset::Utc;
+use hawk::{Error as HawkError, Header as HawkHeader, Key, mac::{Mac as HawkMac, MacType}};
+use hkdf::{Hkdf, InvalidLength};
+use hmac::{
+    crypto_mac::{InvalidKeyLength, MacError},
+    Hmac, Mac,
+};
+use ring;
+use serde_json::{self, Error as JsonError};
+use sha2::Sha256;
+use time::Duration;
+
+use settings::Settings;
+
+#[derive(Debug, Deserialize)]
+pub struct HawkPayload {
+    expires: f64,
+    node: String,
+    salt: String,
+    uid: u64,
+}
+
+impl HawkPayload {
+    fn new(
+        header: &str,
+        method: &str,
+        host: &str,
+        port: u16,
+        path: &str,
+        settings: &Settings,
+        expiry: u64
+    ) -> AuthResult<HawkPayload> {
+        if &header[0..5] != "Hawk " {
+            return Err(AuthError);
+        }
+
+        // Extract id from hawk header
+        let header: HawkHeader = header[5..].parse()?;
+        let id = header.id.as_ref().ok_or(AuthError)?;
+
+        // Extract payload and signature from hawk id
+        let decoded_id = base64::decode_config(id, base64::URL_SAFE)?;
+        let payload_length = decoded_id.len() - 32;
+        let payload = &decoded_id[0..payload_length];
+        let signature = &decoded_id[payload_length..];
+
+        // Derive signing secret from master secret
+        let hkdf: Hkdf<Sha256> = Hkdf::extract(None, &settings.master_secret);
+        let mut signing_secret = [0u8; 32];
+        hkdf.expand(
+            b"services.mozilla.com/tokenlib/v1/signing",
+            &mut signing_secret,
+        )?;
+
+        // Verify signature
+        let mut expected_signature: Hmac<Sha256> = Hmac::new_varkey(&signing_secret)?;
+        expected_signature.input(&payload);
+        expected_signature.verify(signature)?;
+
+        // Parse payload JSON
+        let payload: HawkPayload = serde_json::from_slice(payload)?;
+        println!("{:?}", payload);
+
+        // Check payload is not expired
```

Comment:
FYI in [this commit](https://github.com/mozilla-services/go-syncstorage/commit/ba4f08c668d742754906ef950c91bb9c670bf687)  I made go-syncstorage accept an environment configurable timestamp skew.  In reality some client clocks are off by months and every sync request they sent was 403'd ...  I wound up setting this to 1 year just so they could sync.

Reply:
👍 Thanks for the tip, will do the same here!